lib/bundler/definition.rb

```diff
@@ -113,6 +114,10 @@ def initialize(lockfile, dependencies, sources, unlock, ruby_version = nil, opti
         @unlock[:gems] = @locked_specs.for(eager_unlock).map(&:name)
       end
 
+      if @unlock.is_a?(Hash) && @unlock[:sources]
```

Comment:
Come to think of this line, line 112 with its `unless @unlock[:lock_shared_dependencies]` would have already failed if the `:[]` method was not available.